management/src/main/java/org/kaazing/gateway/management/filter/ManagementFilter.java

```diff
@@ -57,18 +51,11 @@
     protected ServiceContext serviceContext;
     private ServiceCounterManager serviceCounterManager;
 
-    public ManagementFilter(ServiceManagementBean serviceBean,
-                            MonitoringEntityFactory monitoringEntityFactory,
-                            String serviceName,
-                            Properties configuration) {
+    public ManagementFilter(ServiceManagementBean serviceBean) {
         this.serviceBean = serviceBean;
         this.managementContext = serviceBean.getGatewayManagementBean().getManagementContext();
         this.serviceContext = serviceBean.getServiceContext();
-        String gatewayId = InternalSystemProperty.GATEWAY_IDENTIFIER.getProperty(configuration);
-
-        CounterManagerFactory counterFactory = new CounterManagerFactoryImpl();
-        serviceCounterManager = counterFactory.makeServiceCounterManager(monitoringEntityFactory,
-                serviceName, gatewayId);
+        serviceCounterManager = (ServiceCounterManager) serviceContext.getMonitoringFactory();
         serviceCounterManager.initializeSessionCounters();
```

Comment:
There is a single instance of ServiceCounterManager per service but there is a new instance of ManagementFilter per client connection (IoSession). The ServiceCounterManager instance is therefore shared across all Client connections, and must be able to operate correctly when used simultaneously by multiple threads. Looking at the implementation of initializeSessionCounters it should be called only once per instance of ServiceCounterManager, so I don't think you can call it for each new instance of the filter. It needs to be called once only during service startup, One possible solution would be to get rid of the ServiceCounterManager.initializeSessionCounters method altogether and just do the counter initialization logic in the ServiceCounterManagerImpl constructor.